js/ceo.js

```diff
@@ -0,0 +1,503 @@
+'use strict';
+
+//  ---------------------------------------------------------------------------
+
+const Exchange = require ('./base/Exchange');
+const { ExchangeError, ArgumentsRequired, AuthenticationError, OrderNotFound, ExchangeNotAvailable, InvalidOrder, PermissionDenied } = require ('./base/errors');
+
+//  ---------------------------------------------------------------------------
+
+module.exports = class ceo extends Exchange {
+    describe () {
+        return this.deepExtend (super.describe (), {
+            'id': 'ceo',
+            'name': 'CEO',
+            'countries': [ 'CN' ],
+            'rateLimit': 1000,
+            'version': 'v1',
+            'has': {
+                'CORS': false,
+                'createMarketOrder': false,
+                'fetchDepositAddress': true,
+                'fetchOrder': true,
+                'fetchOrders': true,
+                'fetchOpenOrders': true,
+                'fetchOHLCV': true,
+                'fetchTickers': true,
+                'withdraw': false,
+            },
+            'timeframes': {
+                '1m': '1min',
+                '3m': '3min',
+                '5m': '5min',
+                '15m': '15min',
+                '30m': '30min',
+                '1h': '1hour',
+                '2h': '2hour',
+                '4h': '4hour',
+                '6h': '6hour',
+                '12h': '12hour',
+                '1d': '1day',
+                '3d': '3day',
+                '1w': '1week',
+            },
+            'exceptions': {
+                '1001': ExchangeError,
+                '1002': ExchangeError,
+                '1003': AuthenticationError,
+                '1004': ExchangeNotAvailable,
+                '1005': ExchangeNotAvailable,
+                '1006': PermissionDenied,
+                '1007': ExchangeError,
+                '1008': AuthenticationError,
+            },
+            'urls': {
+                'api': 'https://ceo.bi/api',
+                'www': 'https://ceo.bi',
+                'doc': 'https://ceo.bi/api/doc',
+            },
+            'api': {
+                'public': {
```

Comment:
This should be:

```JavaScript
'urls': {
    // 'api': 'https://ceo.bi/api', // wrong
    'api': { // correct
        'public': 'https://ceo.bi/api/market',
        'private': 'https://ceo.bi/api/deal',
    },
},
'api': {
    'public': {
        'get': [
            // 'market/markets' → 'markets'
            'markets',
            'allTicker',
        ]
    }
    'private': {
        'get': [
            // 'deal/accountInfo' → 'accountInfo'
            'accountInfo',
            'order',
            'getChargeRecord',
            ...
        ],
    },
}
```

↑ Then, after you look at it and compare it to the implementation of ZB – you'll notice that it's a clone of the same engine so you don't have to write the entire implementation from scratch, I'd suggest that you derive this class from the zb class. We have derived exchanges, for example, yobit is derived from liqui and it overrides a few methods for exchange-specific details: https://github.com/ccxt/ccxt/blob/master/js/yobit.js

I suggest that we don't duplicate the code and reuse the existing code, if this implementation is derived from ZB.

Reply:
@zed8 let me know if you can rewrite the above to use the zb as the base class.

Reply:
OK,  I am rewriting the code

Reply:
@zed8 thanks a lot!

Reply:
the code has been refactored

Reply:
> the code has been refactored

 Unfortunately, it has been refactored completely wrong... Instead of deriving it from zb, you've decided to override all [implicit methods](https://github.com/ccxt/ccxt/wiki/Manual#implicit-api-methods) for some reason and now the interface to this exchange is not unified: https://github.com/ccxt/ccxt/wiki/Requirements...

Reply:
For some reason you've ignored the above comment: https://github.com/ccxt/ccxt/pull/4895#discussion_r273418678:

> // 'market/markets' → 'markets'

> ```
>  // 'api': 'https://ceo.bi/api', // wrong
>     'api': { // correct
>         'public': 'https://ceo.bi/api/market',
>         'private': 'https://ceo.bi/api/deal',
>     },
> ```

Reply:
I suggest that we revert back to how it was before the refactoring, and start all over again, basically, you need to follow it very precisely:

1. Revert to this stage: https://github.com/ccxt/ccxt/pull/4895/commits/e764361a3bfdee1b7552c7501f7cc8998513debf
2. Add the changes as described in this comment: https://github.com/ccxt/ccxt/pull/4895#discussion_r273418678
3. Test your exchange – it will not work. To make it work, you will have to go through each method and change `this.publicGetMarketMarkets` to `this.publicGetMarkets` and do the same substitution in each method.
4. If you did steps (2) and (3) very precisely and didn't miss anything – after testing again it will work again.
5. Open zb.js.
6. Compare your methods to implementations of the same methods in zb.js
7. If your method looks very similar and does the same thing or less than the corresponding method in zb – you can delete the method in ceo.js, because it is duplicating the functionality that it will inherit from zb anyway.
8. Remove the rest of duplicate methods (the ones that looks almost identical to implementations in zb).
9. In the end the ceo.js file will contain a short `describe()` method and maybe a few methods that really differ from zb, everything else will be stripped out and inherited from zb (mostly).

Let me know if you have difficulties or questions upon any of the above steps. To avoid doing a lot of redundant work, it's easier to ask before applying the edits ) Or, if you don't want to refix this, let us know your plans, we will try to fix this with the help of the community.

Thx for your help anyway, appreciate your involvement!

Reply:
ceo.js fetchMarkets() method have line:  let markets = response['data'];  ze.js no it , how do

Reply:
@zed8 Good question, thx for asking!

Let's compare their actual outputs for `fetchMarkets` side by side:

We take the following two URLs:

- https://ceo.bi/api/market/markets (it redirects me to https://ceohk.bi/api/market/markets, but it doesn't matter)
- http://api.zb.cn/data/v1/markets

Here's how their output looks like:

![ceo-zb-1](https://user-images.githubusercontent.com/1294454/55931139-11987a80-5c2d-11e9-8dca-8c62e0d694f1.jpg)

It's not immediately obvious that these are the same thing, but, if we add some comments, it becomes obvious then:

![ceo-zb-2](https://user-images.githubusercontent.com/1294454/55931143-152c0180-5c2d-11e9-9fee-3b98a8b55adc.jpg)

It's easy to see that ZB is missing some fields, and CEO response structure has +1 level in depth. But this is still the same engine and the format of the response is almost the same, it allows missing fields. So, in order to avoid code duplication, we do should alter the implementation in zb.

1. We split the fetchMarkets in zb into several methods:

```JavaScript
    fetchMarkets (params = {}) {
        const response = await this.publicGetMarkets ();
        return this.parseMarkets (response);
    }

    parseMarkets (markets) {
        const keys = Object.keys (markets);
        const result = [];
        for (let i = 0; i < keys.length; i++) {
            const id = keys[i];
            const market = markets[id];
            const [ baseId, quoteId ] = id.split ('_');
            const base = this.commonCurrencyCode (baseId.toUpperCase ());
            const quote = this.commonCurrencyCode (quoteId.toUpperCase ());
            const symbol = base + '/' + quote;
            const precision = {
                'amount': market['amountScale'],
                'price': market['priceScale'],
            };
            result.push ({
                'id': id,
                'symbol': symbol,
                'baseId': baseId,
                'quoteId': quoteId,
                'base': base,
                'quote': quote,
                'active': true,
                'precision': precision,
                'limits': {
                    'amount': {
                        'min': Math.pow (10, -precision['amount']),
                        'max': undefined,
                    },
                    'price': {
                        'min': Math.pow (10, -precision['price']),
                        'max': undefined,
                    },
                    'cost': {
                        'min': 0,
                        'max': undefined,
                    },
                },
                'info': market,
            });
        }
        return result;
    }
```

2. Then we should alter `parseMarkets` and we should add the safe code to handle the missing fields. We will be using safe methods, as described here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#sanitizing-input-with-safe-methods

```JavaScript
            // ...
            const id = keys[i];
            const market = markets[id];
            const [ baseId, quoteId ] = id.split ('_');
            const base = this.commonCurrencyCode (baseId.toUpperCase ());
            const quote = this.commonCurrencyCode (quoteId.toUpperCase ());
            const symbol = base + '/' + quote;
            const amountPrecision = this.safeInteger (market, 'amountScale'); // ←------------------
            const pricePrecision = this.safeInteger (market, 'priceScale'); // ←------------------
            const precision = {
                'amount': amountPrecision, // ←------------------
                'price': pricePrecision, // ←------------------
            };
            // ...
```

↑ This way we handle the *number vs string* difference in their formats for `priceScale`.

3. Then to the same `parseMarkets` method we add the handling for the `numericId` field (CEO has it, but ZB doesn't),

```JavaScript
            // ...
            const numericId = this.safeInteger (market, 'id');  // ←------------------
            result.push ({
                'id': id,
                'numericId': numericId,  // ←------------------
                'symbol': symbol,
                'baseId': baseId,
                'quoteId': quoteId,
            // ...
```

4. Then in the same `parseMarkets` method we should handle the limits properly in the same manner, using safe methods, and defaulting to the precision. Let me know if you have questions on that...

5. After the edits are done, we add the following implementation of `fetchMarkets` to ceo.js:

```JavaScript
    fetchMarkets (params = {}) {
        const response = await this.publicGetMarkets ();
        const data = this.safeValue (response, 'data', {});
        return this.parseMarkets (data);
    }
```

Reply:
Do you have time to refactor? I am willing to support 3 ETH

Reply:
I will do my best to get to it as soon as I can, but i can't promise on the timing, unfortunately. Hope for your understanding, we have a long backlog of exchanges to integrate with the same type of issues that we have to fix as well. I will update you here as soon as I have an opportunity to take this task. In the meantime, let us know if anything or feel free to reach out to us via email at info@ccxt.trade.

Reply:
You can still build the version of the lib with CEO and use it in the projects by following these instructions: 

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#what-you-need-to-have
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-build--run-tests-on-your-local-machine

Reply:
Ok, I am expecting to hear your good news.